File: packages/ts/server/src/enrichers/compose.ts
import type { RoleEnricher, AuthResult } from "../router/route-builder.js"

/**
 * Composes multiple `RoleEnricher` functions into one.
 *
 * All enrichers run in parallel (`Promise.all`) and their results are merged into
 * a single flat array. Use this when `SyncRouterOptions.roleEnricher` needs to
 * combine several application-level enrichers — for example a custom team-membership
 * enricher together with an entitlement enricher.
 *
 * ```ts
 * import { composeEnrichers, type RoleEnricher } from "@drakkar.software/starfish-server"
 * import { createEntitlementRoleEnricher } from "@drakkar.software/starfish-entitlements"
 *
 * const teamEnricher: RoleEnricher = async (auth, params) =>
 *   (await isTeamMember(auth.identity, params.teamId)) ? ["team-member"] : []
 *
 * const roleEnricher = composeEnrichers(
 *   teamEnricher,
 *   createEntitlementRoleEnricher({ store }),
 * )
 *
 * const router = createSyncRouter({ store, config, roleResolver, roleEnricher })
 * ```
 */
export function composeEnrichers(...enrichers: RoleEnricher[]): RoleEnricher {
  return async function composedEnricher(
    auth: AuthResult,
    params: Record<string, string>,
  ): Promise<string[]> {
    const results = await Promise.all(enrichers.map((e) => e(auth, params)))
    return results.flat()
  }
}

File: packages/python/server/starfish_server/enrichers/compose.py
"""Utility for composing multiple RoleEnrichers into one."""


import asyncio

from starfish_server.router.route_builder import AuthResult, RoleEnricher


def compose_enrichers(*enrichers: RoleEnricher) -> RoleEnricher:
    """Compose multiple :class:`RoleEnricher` functions into one.

    All enrichers run concurrently via :func:`asyncio.gather` and their results
    are merged into a single flat list.  Use this when
    :class:`~starfish_server.router.route_builder.SyncRouterOptions` needs to
    combine several application-level enrichers — for example a custom
    team-membership enricher together with an entitlement enricher.

    ::

        from starfish_server import compose_enrichers
        from starfish_entitlements import create_entitlement_role_enricher

        async def team_enricher(auth, params):
            if await is_team_member(auth.identity, params.get("teamId")):
                return ["team-member"]
            return []

        role_enricher = compose_enrichers(
            team_enricher,
            create_entitlement_role_enricher(EntitlementRoleEnricherOptions(store=store)),
        )

        router = create_sync_router(SyncRouterOptions(
            store=store, config=config, role_resolver=role_resolver,
            role_enricher=role_enricher,
        ))
    """

    async def _composed(auth: AuthResult, params: dict[str, str]) -> list[str]:
        results = await asyncio.gather(*(e(auth, params) for e in enrichers))
        return [role for sub in results for role in sub]

    return _composed
